### PR TITLE
Sqlite dump tool

### LINKED ISSUE #78: Generate an sqlite dump of records
sqlite is a single file relational database that provides a convenient mechanism for sharing data. 

Prototype generation of an sqlite database that contains the isample core record representation. Schema for the database may initially follow the things structure, though will likely be more normalized for efficiency.

### PR DESCRIPTION
**Pull request by dannymandel**
As requested by @hongcui, we'd like an easy to consume dump of our ISB Core records.  We already had https://github.com/isamplesorg/isamples_inabox/issues/78, so it seemed like a good time to take care of two things at once.

Implementation ends up being pretty straightforward -- create a SQLModel and use each row of solr results to create a new model object then save it with SQLAlchemy.

### CODE REVIEW

File: isb_lib/models/isb_core_record.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
+from typing import Optional
+
+from sqlmodel import SQLModel, Field
+
+
```
Comment (dannymandel) [inline isb_lib/models/isb_core_record.py:5]: This ends up being pretty much as you'd expect -- one field == one column in the SQLite schema.

File: scripts/create_sql_lite_dump.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+import isb_lib.core
+from isamples_metadata import Transformer
+from isb_lib.models.isb_core_record import ISBCoreRecord
+from isb_web.isb_solr_query import ISBCoreSolrRecordIterator
+import requests
+from sqlmodel import SQLModel, create_engine, Session
+import logging
+
+# Use this to map between the solr column names and the sqlite column names and control what is
+# written to the dump.
```
Comment (dannymandel) [inline scripts/create_sql_lite_dump.py:15]: Here we declare which fields we want to include in the dump, as well as rename them when appropriate.

File: scripts/create_sql_lite_dump.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    "curation_label": "curation_label",
+    "curation_description": "curation_description",
+    "curation_accessConstraints": "curation_access_constraints",
+    "curation_location": "curation_location",
+    "curation_responsibility": "curation_responsibility",
+    "relatedResource_isb_core_id": "related_resources_isb_core_id",
+    "source": "source"
+}
+
+NUMERIC_COLUMNS = ["produced_by_sampling_site_elevation_in_meters"]
```
Comment (dannymandel) [inline scripts/create_sql_lite_dump.py]: This ended up being the only real non-string column.  Everything else made sense as a string in SQLite.

File: scripts/create_sql_lite_dump.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+
+# Filter out any placeholder values that shouldn't appear in the dump
+def _filtered_value(value):
+    if type(value) is str and value == Transformer.NOT_PROVIDED:
```
Comment (dannymandel) [inline scripts/create_sql_lite_dump.py:61]: On inspection of the data, we had a fair number of "Not Provided" in the solr data.  Filter these out so they don't end up in the sqlite dump.

File: scripts/create_sql_lite_dump.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    "--authority",
+    default=None,
+    help="Which authority to use when selecting and dumping the resolved_content",
+)
+@click_config_file.configuration_option(config_file_name="isb.cfg")
+@click.pass_context
+def main(ctx, db_url, solr_url, verbosity, heart_rate, authority):
+    isb_lib.core.things_main(ctx, db_url, solr_url, verbosity, heart_rate)
+    engine = create_engine(ctx.obj["db_url"], echo=False)
+    SQLModel.metadata.drop_all(engine, tables=[ISBCoreRecord.__table__])
```
Comment (dannymandel) [inline scripts/create_sql_lite_dump.py:95]: Drop the table every time…made sense for how we're using this and certainly made development easier.

Comment (dannymandel) [reply scripts/create_sql_lite_dump.py:95]: Also, explicitly specify the table so we don't pull in any of the other stuff (Things and related indexes).

File: scripts/create_sql_lite_dump.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    rsession = requests.session()
+    iterator = ISBCoreSolrRecordIterator(rsession, authority, batch_size, 0, "id asc")
+    num_records = 0
+    for solr_record in iterator:
+        new_record = ISBCoreRecord()
+        for key, value in SOLR_TO_SQLITE_FIELD_MAPPINGS.items():
+            solr_value = solr_record.get(key)
+            if type(solr_value) is list:
+                filtered_solr_values = [_filtered_value(value) for value in solr_value]
+                solr_value = "; ".join(filtered_solr_values)
```
Comment (dannymandel) [inline scripts/create_sql_lite_dump.py]: For solr lists, do two things:

(1) Remove any embedded "Not Provided" strings
(2) Turn the list into a semicolon separated string.  SQLite doesn't have a list data type.  Not going to worry about trying to do anything goofy like make a join table for these.

@datadavev if you favor a different approach let me know

File: scripts/create_sql_lite_dump.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+            elif type(solr_value) is str:
+                # Don't include placeholder empty values in the dump
+                solr_value = _filtered_value(solr_value)
+            # Key is source column solr name, value is dest column sqlite name in sqlite
+            if solr_value is not None:
+                new_record.__setattr__(value, solr_value)
+            else:
```
Comment (dannymandel) [inline scripts/create_sql_lite_dump.py:124]: This last bit is a workaround for a hack somewhere below me in the stack (not sure if SQLModel or SQLAlchemy, or SQLite) -- basically if we didn't explicitly initialize a value for a column, it ended up using the column metadata definition as the value and the generated SQL exploded.  I didn't do work to try and pinpoint how/why but verified this workaround works.

File: scripts/create_sql_lite_dump.py (hunk 3 above)
Comment (datadavev) [reply scripts/create_sql_lite_dump.py]: Aren't the various `_categories` fields multi-valued? Should these be strings or a list in the resulting ssqlite?

File: scripts/create_sql_lite_dump.py (hunk 4 above)
Comment (datadavev) [reply scripts/create_sql_lite_dump.py:61]: so "Not Provided" becomes a null in sqlite?

File: scripts/create_sql_lite_dump.py (hunk 6 above)
Comment (datadavev) [reply scripts/create_sql_lite_dump.py]: Replace "Not Provided" with nulls / empty. 

Wrt multivalued fields, sqlite does have JSON support, so could use a JSON field type with a list. Semi-colon delimited is a good option also, but perhaps start with a JSON list and see what sorts of barriers that raises wrt to consumption. If we can do lists, then that opens a lot of flexibility.

File: scripts/create_sql_lite_dump.py (hunk 7 above)
Comment (datadavev) [reply scripts/create_sql_lite_dump.py:124]: Make the column nullable? Then it should accept creation without a value

**Review by datadavev — COMMENTED** [review]
A couple suggestions, but otherwise looks good

File: scripts/create_sql_lite_dump.py (hunk 4 above)
Comment (dannymandel) [reply scripts/create_sql_lite_dump.py:61]: It becomes an empty string due to the other weirdness I ran into, but yes, we don't include "Not Provided" in the dump.

File: scripts/create_sql_lite_dump.py (hunk 6 above)
Comment (dannymandel) [reply scripts/create_sql_lite_dump.py]: Yep -- Not Provided is filtered out.  

The multivalued fields are now stored as JSON lists.  The raw column in the table is a string, but SQLAlchemy takes care of converting it to a list as the record is retrieved.

File: scripts/create_sql_lite_dump.py (hunk 7 above)
Comment (dannymandel) [reply scripts/create_sql_lite_dump.py:124]: Has nothing to do with the column being nullable, unfortunately, I think it's a bug elsewhere in the stack.  Interestingly, the bug persisted even after switching over to the SQLAlchemy Core API.

**Pull request merged**